source/extensions/tracers/opentelemetry/opentelemetry_tracer_impl.cc

```diff
@@ -99,17 +150,19 @@ Tracing::SpanPtr Driver::startSpan(const Tracing::Config& config,
   // Get tracer from TLS and start span.
   auto& tracer = tls_slot_ptr_->getTyped<Driver::TlsTracer>().tracer();
   SpanContextExtractor extractor(trace_context);
+  auto span_kind = getSpanKind(config);
+  auto initial_attributes = getInitialAttributes(trace_context, span_kind);
```

Comment:
This means we will create a map at runtime for every request? This will bring pretty heavy overhead. A basic rule in the in Envoy of adding new feature should be `this feature is harmless to the one who won't use it except for at most a if check.`

Reply:
Thanks for your feedback. I will change the code to create the map only if a sampler is configured.

Reply:
I am not sure. But whether the standard of OTel defined the structure of Attributes? Could we use the TraceContext  directly or similar interface to provided these attributes? Ideally, I hope we do the searching only when a attributes or key is acquired explicitly.

Reply:
one of the parameters of `shouldSample` is a "Initial set of Attributes of the Span to be created.". 
According to https://opentelemetry.io/docs/specs/otel/common/#attribute , an attribute is an attribute is a key-value pair.

The sampler interface introduced in #30259 was aligned to the OTel spec in 
https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/trace/sdk.md#shouldsample

So passing the TraceContext to the tracer would not be aligned to the otel tracer. There might also be additional "initialAttributes" which can't be gathered using TraceContext. And I think it should also be possible to pass these attributes to a sampler.

Reply:
But you will still search the `TraceContext` (by `getByKey()`) and populate the attribute map even these attributes never be used by the sampler.

And I didn't get what is the difference between the `getByKey()` of `TraceContext` and `find()` of `std::map` in this scenario.

Reply:
ok, so I will change `Sampler::shouldSample()` to receive a `TraceContext` instead of `std::map<std::string, std::string> initialAttributes`. Can you pls. confirm before I do the code change?

Another question: What do you think about moving the span kind detection from constructor `Span::Span` to `opentelemetry_tracer_impl.cc`. In my opinion, this still makes sense because it avoids passing `const Tracing::Config&config` and `bool downstream_span`.

Reply:
I pushed a949a16d75fbcb995cbe89dd363c4857f19d6588. Now the TraceContext is passed to a sampler to provide access to initial span attributes.
I used `Envoy::OptRef` because a TraceContext is not always available (if a child span is created).

Reply:
> I am not sure. But whether the standard of OTel defined the structure of Attributes

The `TraceContext` in Envoy seems to be "just" a bag of key-values, so I agree we could simply pass that on to sampler. 

But: The naming `TracingContext` is a bit unfortunate though, because it is both a holder for the `traceparent` but also for attributes (I guess via the `RequestHeaderMap`). Nothing addressable, but I just wanted to point that out, since it my understanding from the W3C trace context spec does not include attributes.